### PR TITLE
Create C++ implementation of consistent hashing for distributions.

### PR DESCRIPTION
**Pull request by tcsnfkx**
The details of the algorithm and the original Python implementation can
be found in
https://github.com/world-federation-of-advertisers/virtual_people_examples/blob/main/notebooks/Consistent_Hashing.ipynb.

### CODE REVIEW

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  // Always use Build to get a DistributedConsistentHashing object. Users should
+  // not call the constructor below directly.
+  //
+  // @distribution is represented by a list of (choice id, probability) pairs.
+  // The probabilities will be normalized.
+  //
+  // Returns error status if any of the following happens:
+  //   @distribution is empty.
+  //   Any probability in @distribution is negative.
+  //   Probabilities sum in @distribution is not positive.
```
Comment (wliue) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.h:40]: Maybe -> "Sum of Probabilities".

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+      DistributedConsistentHashing::Build(std::move(distribution));
+  EXPECT_TRUE(hashing_or.ok());
+  std::unique_ptr<DistributedConsistentHashing> hashing =
+      std::move(hashing_or.value());
+  for (int seed = 0; seed < kSeedNumber; ++seed) {
+    EXPECT_EQ(hashing->Hash(std::to_string(seed)), 1);
+  }
+}
+
+TEST(DistributedConsistentHashingTest, TestNormalizeToZero) {
```
Comment (wliue) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc:162]: Maybe -> TestZeroAfterNormalization?

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+      DistributedConsistentHashing::Build(std::move(distribution));
+  EXPECT_TRUE(hashing_or.ok());
+  std::unique_ptr<DistributedConsistentHashing> hashing =
+      std::move(hashing_or.value());
+  for (int seed = 0; seed < kSeedNumber; ++seed) {
+    EXPECT_EQ(hashing->Hash(std::to_string(seed)), 1);
+  }
+}
+
+TEST(DistributedConsistentHashingTest, TestOutputDistribution_NonConsecutive) {
```
Comment (wliue) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc:183]: Maybe -> NonConsecutiveChoiceId?

**Review by wliue — APPROVED** [review]

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+namespace wfa_virtual_people {
+
+absl::StatusOr<std::unique_ptr<DistributedConsistentHashing>>
+DistributedConsistentHashing::Build(
+    std::unique_ptr<std::vector<std::pair<int32_t, double>>> distribution) {
+  if (distribution->empty()) {
+    return absl::InvalidArgumentError("The given distribution is empty.");
+  }
+  double probabilities_sum = 0.0;
+  for (auto it = distribution->begin(); it != distribution->end(); ++it) {
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc:34]: for-each loop here?

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/str_format.h"
+#include "src/farmhash.h"
+
+namespace wfa_virtual_people {
+
+absl::StatusOr<std::unique_ptr<DistributedConsistentHashing>>
+DistributedConsistentHashing::Build(
+    std::unique_ptr<std::vector<std::pair<int32_t, double>>> distribution) {
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc:29]: Please consider using a struct with meaningful names instead of using `std::pair`

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  }
+  // Normalizes the probabilities.
+  for (auto it = distribution->begin(); it != distribution->end(); ++it) {
+    it->second /= probabilities_sum;
+  }
+  return absl::make_unique<DistributedConsistentHashing>(
+      std::move(distribution));
+}
+
+std::string GetFullSeed(const std::string& random_seed, const int32_t choice) {
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc:51]: absl::string_view please

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  return absl::StrFormat("consistent-hashing-%s-%d", random_seed, choice);
+}
+
+inline double FloatHash(const std::string& full_seed) {
+  return static_cast<double>(util::Fingerprint64(full_seed)) /
+         static_cast<double>(std::numeric_limits<uint64_t>::max());
+}
+
+double ComputeXi(
+    const std::string& random_seed,
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc:61]: absl::string_view please

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "wfa/virtual_people/core/model/utils/distributed_consistent_hashing.h"
+
+#include <math.h>
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc:17]: Please include cmath and cstdlib and prefix mathematical functions with "std::". There are some really subtle bugs that can come from math.h.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace wfa_virtual_people {
+namespace {
+
+constexpr int kSeedNumber = 10000;
+
+TEST(DistributedConsistentHashingTest, TestEmptyDistribution) {
+  std::unique_ptr<std::vector<std::pair<int32_t, double>>> distribution =
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc:29]: Please use auto here to avoid duplicating the type name

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+      absl::make_unique<std::vector<std::pair<int32_t, double>>>();
+  auto hashing_or =
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc:31]: just `hashing` -- please don't suffix with `_or`. https://abseil.io/tips/181

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+  std::unique_ptr<std::vector<std::pair<int32_t, double>>> distribution_1 =
+      absl::make_unique<std::vector<std::pair<int32_t, double>>>();
+  distribution_1->emplace_back(std::make_pair(0, 0.4));
+  distribution_1->emplace_back(std::make_pair(1, 0.2));
+  distribution_1->emplace_back(std::make_pair(2, 0.2));
+  distribution_1->emplace_back(std::make_pair(3, 0.2));
+  auto hashing_or_1 =
+      DistributedConsistentHashing::Build(std::move(distribution_1));
+  EXPECT_TRUE(hashing_or_1.ok());
+  std::unique_ptr<DistributedConsistentHashing> hashing_1 =
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc:243]: `ASSERT_OK_AND_ASSIGN` could be useful here.

**Review by efoxepstein — CHANGES_REQUESTED** [review]

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.h (hunk 1 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.h]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc (hunk 2 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc (hunk 3 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc (hunk 8 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc (hunk 5 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc (hunk 4 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc (hunk 6 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc (hunk 7 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc (hunk 9 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc (hunk 10 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc (hunk 11 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+  // 3         0.4                           0.2
+  auto distribution = absl::make_unique<std::vector<DistributionChoice>>();
+  distribution->emplace_back(0, 0.8);
+  distribution->emplace_back(1, 0.4);
+  distribution->emplace_back(2, 0.4);
+  distribution->emplace_back(3, 0.4);
+  ASSERT_OK_AND_ASSIGN(
+      std::unique_ptr<DistributedConsistentHashing> hashing,
+      DistributedConsistentHashing::Build(std::move(distribution)));
+  absl::flat_hash_map<int32_t, int> ouptut_counts = {
```
Comment (mpatel24) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc:109]: This should be 'output_counts' instead of 'ouptut_counts'

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+  // 6         0.2
+  auto distribution = absl::make_unique<std::vector<DistributionChoice>>();
+  distribution->emplace_back(0, 0.4);
+  distribution->emplace_back(2, 0.2);
+  distribution->emplace_back(4, 0.2);
+  distribution->emplace_back(6, 0.2);
+  ASSERT_OK_AND_ASSIGN(
+      std::unique_ptr<DistributedConsistentHashing> hashing,
+      DistributedConsistentHashing::Build(std::move(distribution)));
+  absl::flat_hash_map<int32_t, int> ouptut_counts = {
```
Comment (mpatel24) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc:183]: same comment as above

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc (hunk 12 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc (hunk 13 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc [hunk 14]
```diff
@@ -14,66 +14,72 @@
...
 
 // A C++ version of the Python function ConsistentHashing.hash in
 // https://github.com/world-federation-of-advertisers/virtual_people_examples/blob/main/notebooks/Consistent_Hashing.ipynb
 int32_t DistributedConsistentHashing::Hash(
-    const std::string& random_seed) const {
+    absl::string_view random_seed) const {
   auto it = distribution_->begin();
-  int32_t choice = it->first;
-  double choice_xi = ComputeXi(random_seed, it->first, it->second);
+  int32_t choice = it->choice_id;
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc:77]: Maybe simpler (up to you):

```
absl::optional<int32_t> choice = absl::nullopt;
double choice_xi = std::numeric_limits<double>::max();
for (const DistributionChoice& distribution_choice : *distribution_) {
  double xi = ComputeXi(...);
  if (choice_xi > xi) {
    choice = distribution_choice.choice_id;
    choice_xi = xi;
  }
}
```

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.h [hunk 15]
```diff
@@ -16,9 +16,18 @@
 #define WFA_VIRTUAL_PEOPLE_CORE_MODEL_UTILS_DISTRIBUTED_CONSISTENT_HASHING_H_
 
 #include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
 
 namespace wfa_virtual_people {
 
+struct DistributionChoice {
+  DistributionChoice(int32_t init_choice_id, double init_probability):
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.h:24]: Structs generally don't need a constructor. You can initialize with `DistributionChoice{some_choice_id, some_probability}`.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc [hunk 16]
```diff
@@ -19,37 +19,36 @@
...
 TEST(DistributedConsistentHashingTest, TestEmptyDistribution) {
-  std::unique_ptr<std::vector<std::pair<int32_t, double>>> distribution =
-      absl::make_unique<std::vector<std::pair<int32_t, double>>>();
-  auto hashing_or =
-      DistributedConsistentHashing::Build(std::move(distribution));
-  EXPECT_EQ(hashing_or.status().code(), absl::StatusCode::kInvalidArgument);
+  auto distribution = absl::make_unique<std::vector<DistributionChoice>>();
+  EXPECT_THAT(
+      DistributedConsistentHashing::Build(std::move(distribution)).status(),
+      wfa::StatusIs(absl::StatusCode::kInvalidArgument, ""));
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc:34]: Please add a using-decl for StatusIs.

```
using ::wfa::StatusIs;
```

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+#include "absl/status/statusor.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/string_view.h"
+#include "src/farmhash.h"
+
+namespace wfa_virtual_people {
+
+absl::StatusOr<std::unique_ptr<DistributedConsistentHashing>>
+DistributedConsistentHashing::Build(
+    std::unique_ptr<std::vector<DistributionChoice>> distribution) {
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc:30]: It's not clear to me why this needs to be a unique_ptr. Could you clarify? Vectors are efficiently moveable -- the unique_ptr seems like an unnecessary indirection. It also adds a bug: what if it's nullptr? Then the `distribution->empty()` check below will crash.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+      DistributedConsistentHashing::Build(std::move(distribution)));
+  absl::flat_hash_map<int32_t, int> output_counts = {
+      {0, 0},
+      {1, 0},
+      {2, 0},
+      {3, 0}};
+  for (int seed = 0; seed < kSeedNumber; ++seed) {
+    int32_t output = hashing->Hash(std::to_string(seed));
+    // The output should always be one of [0, 1, 2, 3].
+    EXPECT_TRUE(output_counts.find(output) != output_counts.end());
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc:77]: Should this be `ASSERT_TRUE`?

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+    ++output_counts[output];
+  }
+  // Absolute error more than 2% is very unlikely.
+  EXPECT_NEAR(
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc:81]: I think you can simplify this (and remove the above EXPECT_TRUE on line 77) with:

```
for (auto& [key, value] : output_counts) value /= static_cast<double>(kSeedNumber);
EXPECT_THAT(
  output_counts,
  UnorderedElementsAre(
    Pair(0, DoubleNear(0.4, 0.02)),
    Pair(1, DoubleNear(0.2, 0.02)),
   ..));
```

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+  ASSERT_OK_AND_ASSIGN(
+      std::unique_ptr<DistributedConsistentHashing> hashing,
+      DistributedConsistentHashing::Build(std::move(distribution)));
+  for (int seed = 0; seed < kSeedNumber; ++seed) {
+    EXPECT_EQ(hashing->Hash(std::to_string(seed)), 1);
+  }
+}
+
+TEST(DistributedConsistentHashingTest,
+     TestOutputDistribution_NonConsecutiveChoiceId) {
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc:169]: Please don't use underscores in googletest test names. https://github.com/google/googletest/blob/master/docs/faq.md#why-should-test-suite-names-and-test-names-not-contain-underscore

**Review by efoxepstein — CHANGES_REQUESTED** [review]

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc (hunk 18 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc:77]: This is for testing that the output is one of the input choice ids.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc (hunk 14 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.h (hunk 15 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.h]: Removed.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc (hunk 16 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc (hunk 17 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc (hunk 18 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc]: This is no longer required.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc (hunk 19 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc (hunk 20 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc [hunk 21]
```diff
@@ -27,14 +27,15 @@ namespace wfa_virtual_people {
 
 absl::StatusOr<std::unique_ptr<DistributedConsistentHashing>>
 DistributedConsistentHashing::Build(
-    std::unique_ptr<std::vector<DistributionChoice>> distribution) {
-  if (distribution->empty()) {
+    std::vector<DistributionChoice>&& distribution) {
+  std::vector<DistributionChoice> raw_distribution(distribution);
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc:31]: Why is this necessary?

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc [hunk 22]
```diff
@@ -166,45 +163,39 @@ TEST(DistributedConsistentHashingTest, TestZeroAfterNormalization) {
...
-      {0, 0},
-      {2, 0},
-      {4, 0},
-      {6, 0}};
+  absl::flat_hash_map<int32_t, double> output_counts;
   for (int seed = 0; seed < kSeedNumber; ++seed) {
     int32_t output = hashing->Hash(std::to_string(seed));
-    // The output should always be one of [0, 2, 4, 6].
-    EXPECT_TRUE(output_counts.find(output) != output_counts.end());
+    if (output_counts.find(output) == output_counts.end()) {
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc:185]: I think this whole if-statement is unnecessary. When you do `output_counts[some_key]`, it will insert a new default value at that key if the key does not yet exist. For doubles, this is 0.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc [hunk 23]
```diff
@@ -99,48 +100,43 @@ TEST(DistributedConsistentHashingTest, TestNormalize) {
...
-      {0, 0},
-      {1, 0},
-      {2, 0},
-      {3, 0}};
+  absl::flat_hash_map<int32_t, double> output_counts;
   for (int seed = 0; seed < kSeedNumber; ++seed) {
     int32_t output = hashing->Hash(std::to_string(seed));
-    // The output should always be one of [0, 1, 2, 3].
-    EXPECT_TRUE(output_counts.find(output) != output_counts.end());
+    if (output_counts.find(output) == output_counts.end()) {
```
Comment (efoxepstein) [inline src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc:115]: I think this whole if-statement is unnecessary. When you do `output_counts[some_key]`, it will insert a new default value at that key if the key does not yet exist. For doubles, this is 0.

**Review by efoxepstein — APPROVED** [review]

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc (hunk 21 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc (hunk 22 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc]: Done.

File: src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc (hunk 23 above)
Comment (tcsnfkx) [reply src/main/cc/wfa/virtual_people/core/model/utils/distributed_consistent_hashing_test.cc]: Done.

**Review by efoxepstein — APPROVED** [review]

**Review by my2457 — APPROVED** [review]
